stan/math/prim/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,156 @@
+       * @return Application of the function specified
+       * by F to the specified scalars.
+       */
+      static inline return_t apply(const T1& x, const T2& y) {
+        return F::fun(x, y);
+      }
+    };
+
+    template <typename F, typename T1, typename T2, int R1, int C1,
+              int R2, int C2>
+    struct apply_scalar_binary<F, Eigen::Matrix<T1, R1, C1>,
+                               Eigen::Matrix<T2, R2, C2> > {
+      /**
+       * Return type, which is calculated recursively as a matrix
+       * of the return type of the contained types T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, T1, T2>
+                                ::return_t, R1, C1> return_t;
+
+      /**
+       * Return the result of applying the function defined by the
+       * template parameter F to the specified matrix arguments.
+       *
+       * @param x Matrix to which operation is applied.
+       * @param y Matrix to which operation is applied.
+       * @return Componentwise application of the function specified
+       * by F to the specified matrices.
+       */
+      static inline return_t apply(const Eigen::Matrix<T1, R1, C1>& x,
+                                   const Eigen::Matrix<T2, R2, C2>& y) {
+        check_matching_dims("binary vectorization", "x", x, "y", y);
+        return_t result(x.rows(), x.cols());
+        for (int i = 0; i < x.size(); ++i)
+          result(i) = apply_scalar_binary<F, T1, T2>::apply(x(i), y(i));
+        return result;
+      }
+    };
+
+    template <typename F, typename T1, typename T2, int R1, int C1,
+              int R2, int C2>
+    struct apply_scalar_binary<F, Eigen::Block<Eigen::Matrix<T1, R1, C1> >,
+                               Eigen::Block<Eigen::Matrix<T2, R2, C2> > > {
+      /**
+       * Return type, which is calculated recursively as an Eigen::Matrix
+       * of the return type of the contained types T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, T1, T2>
+                                ::return_t, R1, C1> return_t;
+
+      /**
+       * Apply the function specified by F elementwise to the
+       * specified argument. This is defined recursively through this
+       * class by passing the arguments as matrices to the
+       * <code>Eigen::Matrix</code> specialization.
+       *
+       * @param x <code>Eigen::Block</code> argument container.
+       * @param y <code>Eigen::Block argument</code> container.
+       * @return Elementwise application of F to the elements of the
+       * containers.
+       */
+      static inline return_t apply(
+          const Eigen::Block<Eigen::Matrix<T1, R1, C1> >& x,
+          const Eigen::Block<Eigen::Matrix<T2, R2, C2> >& y) {
+        return apply_scalar_binary<F, Eigen::Matrix<T1, R1, C1>,
+                                   Eigen::Matrix<T2, R2, C2> >::apply(
+            static_cast<Eigen::Matrix<T1, R1, C1> >(x),
```

Comment:
Why is this copying?  You can just work on the mapped elements directly and avoid the copy.  This should be changed.  At most, you'll have to reimplement the loop to avoid the copy.

Reply:
I did it this way because I thought it would be better not to re-write code and in the unary vectorization case, it gets mapped back to Eigen::Matrix. I can re-write the for loop if it's faster. Should the return type be Eigen::Matrix or Eigen::Block?

Reply:
Yes, it should be faster to just loop over the mapped matrix than do a copy.  And hopefully it's only a one-dimensional loop on the size rather than a doubly-indexed one.

Reply:
I just tried changing it, but it wouldn't compile. `check_matching_dims` expects a `Eigen::Matrix` argument, not a `Eigen::Block` argument.